storage/innobase/handler/ha_innodb.cc

```diff
@@ -18378,9 +18378,21 @@ wait_background_drop_list_empty(THD*, st_mysql_sys_var*, void*, const void*)
 Force innodb to checkpoint. */
 static
 void
-checkpoint_now_set(THD*, st_mysql_sys_var*, void*, const void* save)
+checkpoint_now_set(THD* thd, st_mysql_sys_var*, void*, const void* save)
 {
 	if (*(my_bool*) save) {
+
+		if (srv_read_only_mode) {
+			push_warning_printf(
+			  thd, Sql_condition::WARN_LEVEL_WARN,
+			  HA_ERR_UNSUPPORTED,
+			  "InnoDB doesn't force checkpoint when %s",
+			  (srv_force_recovery == SRV_FORCE_NO_LOG_REDO)
+			  ? "innodb-force-recovery=6."
+			  : "innodb-read-only=1.");
+			return;
+		}
```

Comment:
This is a suggestion for your consideration. We might want to avoid the if condition nesting as follows. 
```
if (!*(my_bool*) save)
  return;

if (srv_read_only_mode) {
  ...
  return;
}
...